lib/codecs/src/decoding/format/avro.rs

```diff
@@ -0,0 +1,302 @@
+        }
+        Ok(log)
+    }
+}
+
+impl Deserializer for AvroDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let log = self.parse_single(bytes, log_namespace)?;
+        Ok(smallvec![log.into()])
+    }
+}
+
+// can't use std::convert::TryFrom because of orphan rules
+pub fn try_from(value: AvroValue) -> vector_common::Result<VrlValue> {
+    // very similar to avro to json see `impl std::convert::TryFrom<AvroValue> for serde_json::Value`
+    // LogEvent has native support for bytes, so it is used for Bytes and Fixed
+    match value {
+        AvroValue::Array(array) => {
+            let mut vector = Vec::new();
+            for item in array {
+                vector.push(try_from(item)?);
+            }
+            Ok(VrlValue::Array(vector))
+        }
+        AvroValue::Boolean(boolean) => Ok(VrlValue::from(boolean)),
+        AvroValue::Bytes(bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Date(d) => Ok(VrlValue::from(d)),
+        AvroValue::Decimal(ref d) => Ok(<Vec<u8>>::try_from(d)
+            .map(|vec| VrlValue::Array(vec.into_iter().map(VrlValue::from).collect()))?),
+        AvroValue::Double(double) => Ok(VrlValue::from(double)),
+        AvroValue::Duration(d) => Ok(VrlValue::Array(
+            <[u8; 12]>::from(d)
+                .into_iter()
+                .map(VrlValue::from)
+                .collect(),
+        )),
+        AvroValue::Enum(_, string) => Ok(VrlValue::from(string)),
+        AvroValue::Fixed(_, bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Float(float) => Ok(VrlValue::from(float as f64)),
+        AvroValue::Int(int) => Ok(VrlValue::from(int)),
+        AvroValue::Long(long) => Ok(VrlValue::from(long)),
+        AvroValue::Map(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::Null => Ok(VrlValue::Null),
+        AvroValue::Record(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::String(string) => Ok(VrlValue::from(string)),
+        AvroValue::TimeMicros(time_micros) => Ok(VrlValue::from(time_micros)),
+        AvroValue::TimeMillis(time_millis) => Ok(VrlValue::from(time_millis)),
+        AvroValue::TimestampMicros(timestamp_micros) => Ok(VrlValue::from(timestamp_micros)),
+        AvroValue::TimestampMillis(timestamp_millis) => Ok(VrlValue::from(timestamp_millis)),
+        AvroValue::Union(_, v) => try_from(*v),
+        AvroValue::Uuid(uuid) => Ok(VrlValue::from(uuid.as_hyphenated().to_string())),
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
💬 Can we also add one simple roundtrip test? You can look at `lib/codecs/tests/native.rs` for inspiration.

Reply:
I pushed a commit to fix timestamp conversions. 

Also, I noticed that `AvroValue::TimeMicros` and `AvroValue::TimeMillis` are converted to integers although the appropriate conversion would be `Duration`? As long as we can encode/decode this correctly it's fine to keep it as is. That further increases the need for a roundtrip test.

Reply:
Here it is intended to convert AVRO schema definitions to vectors for checksumming by other modules similar to what syslog does. However others decoders like protobuf seem to be simple, so I removed the todo tag here by 
[9aeab12.](https://github.com/vectordotdev/vector/pull/18636/commits/9ff71550c4b9c087a56ccc01e3280aeadca889bc)

Reply:
Thanks for your help!
Letting `AvroValue::TimeMicros` and `AvroValue::TimeMillis` be `Duration` needs to introduce a new type for `Vrl` which may be a lots of work, I think that keeping it as original is better.

Reply:
Do you have some suggest for roundtrip test? I looked `lib/codecs/tests/native.rs` and found that it use `quickcheck::Gen` to  generate test data and I didn't found how to generate the schema definition file. Could I do it just like `lib/codecs/tests/protobuf.rs` which use a simple roundtrip test case ?

Reply:
Hi @WarmSnowy, thank you for these updates!

You can look at this example [here](https://github.com/vectordotdev/vector/blob/master/lib/codecs/tests/native_json.rs#L10) instead. Let me know if there are any follow-up questions.

Reply:
I submitted some commits to do roundtrip test.
Some avro type test items were omitted due to some detail differences like `i32`.

Reply:
You can add all those new strings to `allow.txt`.